### PR TITLE
fix bad LabelKeyMatcher

### LINKED ISSUE #389: kubernetes network policy not work when namespace label contains '-'
network policy like that:

```
apiVersion: v1
items:
- apiVersion: extensions/v1beta1
  kind: NetworkPolicy
  metadata:
    creationTimestamp: 2017-04-13T06:09:25Z
    generation: 1
    name: "3"
    namespace: "3"
    resourceVersion: "716035"
    selfLink: /apis/extensions/v1beta1/namespaces/3/networkpolicies/3
    uid: bef69619-200f-11e7-a4d0-5254008350a3
  spec:
    ingress:
    - from:
      - namespaceSelector:
          matchLabels:
            kubernetes-admin.caicloud.io/partition: "4"
    podSelector: {}
kind: List
metadata: {}
resourceVersion: ""
selfLink: ""
```

Calico node log:

```
2017-04-13 06:35:09.740 [INFO][31] int_dataplane.go 554: Finished applying updates to dataplane. msecToApply=18.040078
2017-04-13 06:35:12.478 [WARNING][31] validation_filter.go 57: Validation failed; treating as missing error=error with field src_selector = 'k8s_ns/label/kubernetes-admin.caicloud.io/partition == '4'' key=Policy(name=np.projectcalico.org/3.3) value=order:1000,selector:"calico/k8s_ns == '3'",inbound:allow from selector "k8s_ns/label/kubernetes-admin.caicloud.io/partition == '4'",outbound:,untracked:false
2017-04-13 06:35:12.478 [INFO][31] label_index.go 67: Deleting selector Policy(name=np.projectcalico.org/3.3)
2017-04-13 06:35:17.980 [INFO][31] int_dataplane.go 536: Applying dataplane updates
```

### PR DESCRIPTION
**Pull request by ijumps**
fixes #389

### CODE REVIEW

File: lib/selector/tokenizer/tokenizer.go [hunk 1]
```diff
@@ -55,7 +55,7 @@ type Token struct {
 
 const (
 	// LabelKeyMatcher is the base regex for a valid label key.
-	LabelKeyMatcher = `([a-zA-Z0-9_.-/]{0,253}/)?[a-zA-Z0-9]([a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])?`
+	LabelKeyMatcher = `([a-zA-Z0-9_.\-/]{0,253}/)?[a-zA-Z0-9]([a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])?`
```
Comment (tmjd) [inline lib/selector/tokenizer/tokenizer.go]: Should that 3rd character class be updated too?
`([a-zA-Z0-9_.\-/]{0,253}/)?[a-zA-Z0-9]([a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])?` <= from
`([a-zA-Z0-9_.\-/]{0,253}/)?[a-zA-Z0-9]([a-zA-Z0-9_.\-]{0,61}[a-zA-Z0-9])?` <= to this

Comment (ijumps) [inline lib/selector/tokenizer/tokenizer.go]: It's not necessary since `-` is the last char

**Comment by ijumps** [comment]
@tmjd Hi, I found it's hard to build a calico node image with this PR, I try to update calico node vendor with a bit hack ( casue my fork address https://github.com/ijumps/libcalico-go/ diff from https://github.com/projectcalico/libcalico-go, I had to replace the file in vendor directory), this works and I build a new calico node images.

But the new images do not work, I finally found I need to rebuild felix , this is not easy (I'm not able to release a new felix images, or I had to change the makefile and do more trick, or I found a wrong way to do this).

Could you help me to build / tell me how to build an image with this PR ? I need to fix this. Thank you very much!

**Comment by tmjd** [comment]
I can try to help but will admit that I have not rebuilt this whole chain like you are trying to do.  I think these are the steps you would need to do:

1. Have a go environment with the following repos checked out under src/github.com/projectcalico.  Note these must be at the specified path for the imports to work, you can have your fork with your changes in those locations.
    - libcalico-go
    - felix
    - calicoctl
2. In the felix repo remove from its `glide.lock` the section with `name: github.com/projectcalico/libcalico-go`.
3. Run `make clean`
4. Run `make vendor`
5. Run `LIBCALICOGO_PATH=<path to libcalicogo> make bin/calico-felix`
6. Copy `bin/calico-felix` into the repo `projectcalico/calicoctl` at the path `calico_node/filesystem/bin`
7. In the calicoctl repo run `make calico/node`

If you need it you could also do the same glide.lock change and rebuild calicoctl with your libcalico-go changes but I don't think you need that.

**Comment by ijumps** [comment]
Hi, I do a little trick by:

1. comment all `glide install` in makefile, and run `glide install` local, then replace the libcalico-go directory in vendor with this PR.
2. build felix normally, and tag the images
3. comment `docker rm` in calicoctl makefile, change the felix name, and build calico node normally

Finally I build an image which solves the issue here.

Thank you for your help, I'll try your solution later.

**Comment by robbrockbank** [comment]
LGTM

**Comment by robbrockbank** [comment]
Thanks @ijumps - we should be rolling this into an RC this week.

**Pull request merged**